### ISSUE TITLE
Support multiple discriminated unions on same object

### ISSUE THREAD

**Issue by Aaronius** [colinhacks/zod#1015]
First of all, thank you for Zod! I'm just starting to use it on a new project. I have a case come up quite frequently that I don't know how to resolve using current Zod functionality. I've asked a similar question [here](https://github.com/colinhacks/zod/discussions/1006) and haven't received a response, so I'm logging an issue considering functionality may need to be added to Zod.

I believe I want multiple discriminated unions on the same object. Here's some code that illustrates what I'm trying to accomplish:

```js
const schema = z.object({
    sku: z.string()
  }).merge(z.discriminatedUnion("product", [
    z.object({
      product: z.literal("shoes"),
      price: z.number(),
      material: z.union([z.literal("leather"), z.literal("suede")])
    }),
    z.object({
      product: z.literal("shirt"),
      price: z.string()
    })
 ]).merge(z.discriminatedUnion("color", [
    z.object({
       color: z.literal("red"),
       size: z.enum(["s", "m"])
    }),
    z.object({
      color: z.literal("blue"),
      size: z.enum(["m", "l"])
    })
  ]);
```

In other words, the object will (1) always have a sku, (2) always have `product` and `price` (and sometime `material`), and (3) always have `color` and `size`.  The combinations would be restricted by the discriminated unions.

This doesn't work, though, because ZodDiscriminatedUnion doesn't have a merge property and ZodDiscriminatedUnion can't be merged into ZodObject.

I'm not particular about the implementation, so long as it's simple (it's a common use case) and preferably does not require `refine` or `superRefine` (though I'd still like to understand how solve it with `refine` or `superRefine` in the meantime). Thanks!

**Comment by FlorianWendelborn** [comment]
Not a fix on zod’s end, but it’s often a good idea to split this up into sub-objects like this:

```ts
const schema = z.object({
	sku: z.string(),
	product: z.discriminatedUnion('type', [
		z.object({
			type: z.literal('shoes'),
			price: z.number(),
			material: z.union([z.literal('leather'), z.literal('suede')]),
		}),
		z.object({
			type: z.literal('shirt'),
			price: z.string(),
		}),
	]),
	color: z.discriminatedUnion('type', [
		z.object({
			type: z.literal('red'),
			size: z.enum(['s', 'm']),
		}),
		z.object({
			type: z.literal('blue'),
			size: z.enum(['m', 'l']),
		}),
	]),
})
```

Obviously, for `color` it doesn’t make much semantic sense to name it `type`, but you get the idea.

As a general feedback, this seems a bit like overengineering. Are you sure you’re actually benefiting from enforcing this? Would it be okay to not be as strict? Available sizes etc. for products usually belong in the database layer, not really into zod validation. You can enforce the restrictions you want via SQL constraints and good table schemas. For example, a combination of color and size could be its own row in a table, with its own ID. You can then enforce that only valid combinations exist by making `color` on the object from your example a foreign key relationship to that table.

**Comment by scotttrinh** [comment]
I'm actually not against bringing some (all?) of ZodObject's methods to `ZodDiscriminatedUnion` since it is in some senses a subtype, but I'm not sure what the level of effort there is off the top of my head. Would be a nice thing for someone to investigate!

**Comment by Aaronius** [comment]
Thanks @FlorianWendelborn. I'd rather not adjust my model to fit Zod's current capabilities if at all possible. 

> Are you sure you’re actually benefiting from enforcing this?

Yes.

> Would it be okay to not be as strict?

Not in this case.

> Available sizes etc. for products usually belong in the database layer, not really into zod validation.

My real world problem deals with taxes rather than products. The products example was just a simplified version of what I'm dealing with. Either way, I'm presenting a form with these inputs and need to validate the user's input combinations on the client before validation occurs on the server.

@scotttrinh That would be excellent. 🙏

**Comment by Aaronius** [comment]
> I'm actually not against bringing some (all?) of ZodObject's methods to `ZodDiscriminatedUnion` since it is in some senses a subtype, but I'm not sure what the level of effort there is off the top of my head. Would be a nice thing for someone to investigate!

I noticed it would also be handy to have the `passthrough` method on `ZodDiscriminatedUnion` as well.

**Comment by whengely** [comment]
I also run into this issue frequently.  Allowing something like the following would be very helpful.

`const firstStepBase = z.object({
  step: z.literal('first'),
})

const custom = z
  .object({
    type: z.literal('custom'),
    customProperty: z.string(),
  })
  .merge(firstStepBase)

const nonCustom = z
  .object({
    type: z.literal('nonCustom'),
    nonCustomProperty: z.number(),
  })
  .merge(firstStepBase)

const firstStep = z.discriminatedUnion('type', [custom, nonCustom])

const secondStep = z.object({
  step: z.literal('second'),
  newProperty: z.enum(['hello', 'there']),
})

z.discriminatedUnion('step', [firstStep, secondStep])` //firstStep errors

**Comment by Aaronius** [comment]
Bump.

**Comment by Aaronius** [comment]
Hm...I added further activity, but the issue was still auto-closed.

**Comment by FlorianWendelborn** [comment]
@Aaronius as the issue author you should be able to re-open it, I think?

**Comment by Aaronius** [comment]
I can't or I would. I don't have the necessary rights.

**Comment by perry-mitchell** [comment]
Having this issue now too. I have a large object schema that has 2 properties that I want to merge using a discriminated union, such that for each value `a`, `b` might be a string, an array of strings or null.

Eg.

```
const MetaSchema = z
    .object({
        // <snip> - 20+ properties etc.
    })
    .merge(z.discriminatedUnion("name", [
        z.object({
            name: z.literal(null),
            placeholder: z.string().optional()
        }),
        z.object({
            name: z.literal(NamePromptType.FullName),
            placeholder: z.string().nullable().optional()
        }),
        z.object({
            name: z.literal(NamePromptType.FirstName),
            placeholder: z.string().nullable().optional()
        }),
        z.object({
            name: z.literal(NamePromptType.LastName),
            placeholder: z.string().nullable().optional()
        }),
        z.object({
            name: z.literal(NamePromptType.Separate),
            placeholder: z.tuple([z.string().nullable(), z.string().nullable()]).optional()
        })
    ]))
    .catchall(z.unknown());
```

While I've seen other suggestions for abstracting the base `MetaSchema` out and then passing that in as a merge in each union value, that falls apart imo when there will be other union'd parts of the `MetaSchema` I might break out next.

**Comment by pointlessrapunzel** [comment]
> how can I do this if there's 2+ discriminated unions being merged in the same object

Having exactly this issue right now.
We're making a branching form, where there might be different fields depending on the values of certain keys and we want to make one final schema to validate it all in the end. Maybe there's an issue with schema design on our end, but we couldn't find a solution yet...

**Comment by Hobbit44** [comment]
I have ran into this too.

@FlorianWendelborn Could this be reopened, please?

**Comment by FlorianWendelborn** [comment]
> @FlorianWendelborn Could this be reopened, please?

I sadly do not possess the power to wield the reopen button. I think you should open a separate issue and reference this one if applicable. Good luck
